datafusion/core/src/physical_plan/sorts/cursor.rs

```diff
@@ -93,3 +96,232 @@ impl Cursor for RowCursor {
         t
     }
 }
+
+/// A cursor over sorted, nullable [`ArrowNativeTypeOp`]
+///
+/// Note: comparing cursors with different `SortOptions` will yield an arbitrary ordering
+#[derive(Debug)]
+pub struct PrimitiveCursor<T: ArrowNativeTypeOp> {
+    values: ScalarBuffer<T>,
+    offset: usize,
+    // If nulls first, the first non-null index
+    // Otherwise, the first null index
+    null_threshold: usize,
```

Comment:
I'm quite pleased with this formulation of nulls, it avoids needing to consult the null mask at all :smile:

Reply:
Is there any reason it is called `null_threshold` rather than `null_index` given it is a null index 🤔

Reply:
I thought null_index might be confused for the index of the first null, which is not always the case

Reply:
that makes sense.